### ISSUE TITLE
Pretrained JF17K- Self-Adversarial

### ISSUE THREAD

**Issue by Finnhh** [ralphabb/BoxE#2]
Hi, Ralph.

First of all, thanks for your contributions. It's really an interesting work.

However, when I use this command to train **JF17K- Self-Adversarial** as follows, I found that it takes **5 days** to train.
`python Training.py "JF17K" -epochs 1000 -validCkpt 100 -learningRate 0.0001 -nbNegExp 100 -lossMargin 5 -embDim 200 -batchSize 1024 -negSampling selfadv -advTemp 2.0 -useTB F`

So I would like to ask if you can give me or upload pretrained files for JF17K- Self-Adversarial. I will be very grateful.

Or would you like to provide the following data for JF17K- Self-Adversarial:

arity | MRR | Hit@1 | Hit@3 | Hit@10 | MR
-- | -- | -- | -- | -- | --
2 |   |   |   |   |  
3 |   |   |   |   |  
4 |   |   |   |   |  
5 |   |   |   |   |  
6 |   |   |   |   |  

Sorry to disturb you and look forward to your reply.

**Comment by ralphabb** [comment]
Hi, 
Thanks for your interest!
In our setup, training takes roughly 10-15 seconds per epoch (Server with V100 GPU, 64 GB of RAM
). So could you let me know your setup, the speed per epoch as shown in the training log, and validation time? 

I've also uploaded pre-trained weights for you: 
https://drive.google.com/drive/folders/1u-kquFGpgMdrHQxxgeDm3hatwAezSrPn?usp=sharing 

All the best,
Ralph

**Comment by Finnhh** [comment]
Hi, @ralphabb 

Thank you for sharing pre-trained weights.

I used CPU (_Intel(R) Xeon(R) Silver 4214 CPU @ 2.20GHz_) to run this code. So it's very slow.

I tried running on a GPU (RTX 2080Ti, 11GB Memory, CUDA Version: 10.2), but it didn't run. I installed the packages using the following command: 
`conda install tensorflow-gpu=1.15.0 cudatoolkit=10.2`
`pip install msgpack==0.6.1`
`pip install msgpack-numpy==0.4.4.2`

I don’t know if there is any problem with this environment.

All the best,
Finnhh

**Comment by ralphabb** [comment]
I suspect it would be something to do with the environment, indeed. It could be the CUDA version. My setup uses CUDA 9.2.

**Comment by Finnhh** [comment]
Hi, @ralphabb

I used the pre-training weights of JF17K you shared, but got these results.
```
MR:14691.876406140145
MRR:0.0003825746690191402
Hits@1:5.858917272088118e-05
Hits@3:0.00014647293180220295
Hits@5:0.00020506210452308414
Hits@10:0.0003075931567846262
```
Here is my file directory, I put your shared files in the BoxE folder.
![image](https://user-images.githubusercontent.com/20140308/109585537-a9175700-7b3e-11eb-9eee-44ea2c40e5f5.png)

I tested JF17K according to arity, the results are as follows:
arity | MRR | Hit@1 | Hit@3 | Hit@10 | MR
-- | -- | -- | -- | -- | --
2 | 0.0004 | 0.0000465 | 0.000093 | 0.00033 | 14698
3 | 0.0003 | 0.00006 | 0.00006 | 0.0002 | 15004
4 | 0 | 0 | 0 | 0 | 14593
5 | 0.002 | 0.001 | 0.001 | 0.003 | 14567
6 | 0 | 0 | 0 | 0 | 20221
all | 0.004 | 0.000059 | 0.00015 | 0.00031 | 14692

Here are JF17K files, I divided the test file by arity:
https://drive.google.com/drive/folders/1SZNOaKOuyILDatwP5qDiv77x6B4dIx-9?usp=sharing

When I used weights_FB-AUTO to test FB-AUTO, I got the same good results as in the paper:
arity | MRR | Hit@1 | Hit@3 | Hit@10 | MR
-- | -- | -- | -- | -- | --
2 | 0.48 | 0.418 | 0.519 | 0.6 | 418
4 | 0.473 | 0.415 | 0.511 | 0.568 | 706
5 | 0.922 | 0.896 | 0.94 | 0.972 | 40
all | 0.834 | 0.801 | 0.856 | 0.897 | 122

I don't know why I can't get similar results on JF17K as in the paper, look forward to your help.

All the best,
Finnhh

**Comment by ralphabb** [comment]
Hi Finnhh, 
The JF17K numbers you are reporting seem to be random, suggesting that the model has failed to load the weights, and thus initialized randomly. Could you check whether the model is successfully loading the weights? A probable reason why it would not is a discrepancy in settings.

**Comment by Finnhh** [comment]
Hi, @ralphabb,

You are right. I just debugged it. Then I found that function _load_params_ in BoxEModel.py reported an error and initialized it randomly.
```
def load_params(self, param_loc=None):
        if self.saver is None:
            self.saver = tf.train.Saver(name="Saver")
        self.sess = tf.Session()
        if param_loc is None:
            param_loc = self.param_directory
        try:
            self.saver.restore(self.sess, param_loc)
            self.sess.run(tf.tables_initializer())
        except Exception as e:
            print(e)
            self.sess.run(tf.global_variables_initializer())
            self.sess.run(tf.tables_initializer())
```
Errors are reported as follows:
```
W tensorflow/core/framework/op_kernel.cc:1651] OP_REQUIRES failed at save_restore_v2_ops.cc:184 : Not found: Key Variable not found in checkpoint
Restoring from checkpoint failed. This is most likely due to a Variable name or other graph key that is missing from the checkpoint. Please ensure that you have not altered the graph expected based on the checkpoint. Original error:

Key Variable not found in checkpoint
         [[node Saver/RestoreV2 (defined at /anaconda3/envs/boxe/lib/python3.7/site-packages/tensorflow_core/python/framework/ops.py:1748) ]]
```

In fact, the function read the correct _param_loc_. But when it ran the next command _self.saver.restore(self.sess, param_loc)_, it couldn't find the correct checkpoint. I don't understand why it can't restore checkpoint.

**Comment by ralphabb** [comment]
Ok, then can you please tell me which ops aren't being restored? I suspect this is due to either 1) a version mismatch between the latest BoxE and the version with which the pre-trained weights were produced, or most likely 2) a discrepancy between the settings you used and those used for producing the weights. Could you share the settings you are using?

**Comment by Finnhh** [comment]
Hi, @ralphabb,

My setting is as follows, which is the same as the parameters given in _ReadMe.md_.
`python Training.py "JF17K" -epochs 1000 -validCkpt 100 -learningRate 0.0001 -nbNegExp 100 -lossMargin 5 -embDim 200 -batchSize 1024 -negSampling selfadv -advTemp 2.0 -useTB F`

I suspect that our environment configuration is different, which makes it impossible to read the weights.

So I reinstalled the environment, the command is as follows (share it with those who may have the same problem):
```
conda install tensorflow-gpu==1.15
conda install cudatoolkit==9.2
pip install msgpack==0.6.1
pip install msgpack-numpy==0.4.4.2
```
It worked! Then I trained the model, the weights of JF17K is as follows:
https://drive.google.com/drive/folders/1SZNOaKOuyILDatwP5qDiv77x6B4dIx-9?usp=sharing

Your work is really amazing. In addition, I have always felt very embarrassed to disturb you. And You have been very patient in answering my questions. I really admire you. I also want to be a person who is willing to share like you. Thank you so much!

All the best,
Finnhh
